src/lib/interpolate.ml

```diff
@@ -0,0 +1,223 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*)
+
+open Util
+
+let linear (ax, ay) (bx, by) =
+  if ax = bx then
+    invalidArg "equal x values %f" ax
+  else
+    let slope = Float.((by - ay) / (bx - ax)) in
+    fun x -> Float.(slope * (x - ax) + ay)
+
+module Tri_Diagonal = struct
+
+  let mult m x =
+    let n = Array.length m in
+    Array.init n (fun i ->
+      let a, b, c = m.(i) in
+      if i = 0 then
+        b *. x.(0) +. c *. x.(1)
+      else if i = n - 1 then
+        a *. x.(i-1) +. b *. x.(i)
+      else
+        a *. x.(i-1) +. b *. x.(i) +. c *. x.(i+1))
+
+  (* Solve a system of equations of the form
+    a_i*x_{i-1} + b_i*x_i + c_i*x_{i+1} = d_i, where a_1 = c_n = 0.
+    The first arr of triples is of (a_i, b_i, c_i) and a_0 and c_n are ignored.
+    The second arr of d.
+
+    This probably has uses outside of spline interpolation.
+    TODO: Now that we're linking against Lacaml, makes sense to test if using
+    lapack improves performance or precision.*)
+  let solve r d =
+    let a i = fst3 (r.(i)) in
+    let b i = snd3 (r.(i)) in
+    let c i = thr3 (r.(i)) in
+    let n = Array.length r in
+    if n = 1 then
+      [| d.(0) /. (b 0) |]
+    else
+      let nm1 = n - 1 in
+      let cm = Array.make nm1 nan in
+      let dm = Array.make nm1 nan in
+      for i = 0 to n - 2 do
+        if i = 0 then begin
+          cm.(i) <- (c i) /. (b i);
+          dm.(i) <- d.(i) /. (b i)
+        end else begin
+          let im1 = i - 1 in
+          cm.(i) <- (c i) /. ((b i) -. cm.(im1) *. (a i));
+          dm.(i) <- (d.(i) -. dm.(im1) *. (a i)) /. ((b i) -. cm.(im1) *. (a i))
+        end
+      done;
+      let x = Array.make n nan in
+      x.(nm1) <- (d.(nm1) -. dm.(nm1-1) *. (a nm1))
+                /. ((b nm1)-.(cm.(nm1-1))*.(a nm1));
+      for i = n - 2 downto 0 do
+        x.(i) <- dm.(i) -. cm.(i) *. x.(i+1)
+      done;
+      x
+ 
+end (* Tri_Diagonal *)
+ 
+module Spline = struct
+
+  open Tri_Diagonal
+ 
+  type boundary_condition =
+    | Natural
+    | Clamped of float * float
+
+  type t = (float * float * (float * float * float) option) array
+
+  let knots = Array.map (fun (x,y,_) -> (x,y))
+
+  let coefficients t =
+    Array.init (Array.length t - 1)
+      (fun i ->
+        let x, y, opt = t.(i) in
+        match opt with
+        | None -> invalidArg "Improper spline size %d" i
+        | Some (c, b, a)  -> y, c, b, a)
+
+  (* Helper functions to extract correctly indexed elements of data.
+     TODO: is this currying worthwhile? Does this make the code more
+     or less readable. *)
+  let h arr i =
+    fst arr.(i + 1) -. fst arr.(i)
+
+  let s arr =
```

Comment:
I know this is probably ocamly, because it's definitely how functional programmers tend to name things (me!), but I don't really get why they're called s & h?